orquesta/utils/specs.py

```diff
@@ -16,18 +16,68 @@
 import six
 import yaml
 
+try:
+    from yaml import CLoader as Loader
+except ImportError:
+    from yaml import Loader
+
+from yaml import constructor
+from yaml import nodes
+
+from orquesta import exceptions as exc
 from orquesta.specs import loader as spec_loader
 
 
 LOG = logging.getLogger(__name__)
 
 
+# Custom YAML loader that throw an exception on duplicate key.
+# Credit: https://gist.github.com/pypt/94d747fe5180851196eb
+class UniqueKeyLoader(Loader):
+    def construct_mapping(self, node, deep=False):
+        if not isinstance(node, nodes.MappingNode):
+            raise constructor.ConstructorError(
+                None, None, "expected a mapping node, but found %s" % node.id, node.start_mark
+            )
+        mapping = {}
+        for key_node, value_node in node.value:
+            key = self.construct_object(key_node, deep=deep)
+            try:
+                hash(key)
+            except TypeError as exc:
+                raise constructor.ConstructorError(
+                    "while constructing a mapping",
+                    node.start_mark,
+                    "found unacceptable key (%s)" % exc,
+                    key_node.start_mark,
+                )
+            # check for duplicate keys
+            if key in mapping:
+                raise constructor.ConstructorError('found duplicate key "%s"' % key_node.value)
+            value = self.construct_object(value_node, deep=deep)
+            mapping[key] = value
+        return mapping
+
+
+# Add UniqueKeyLoader to the yaml SafeLoader so it is invoked by safe_load.
+yaml.add_constructor(
+    yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
+    UniqueKeyLoader.construct_mapping,
+    Loader=yaml.SafeLoader,
+)
+
+
 def instantiate(spec_type, definition):
     if not definition:
         raise ValueError("Workflow definition is empty.")
 
     if isinstance(definition, six.string_types):
-        definition = yaml.safe_load(definition)
+        try:
+            definition = yaml.safe_load(definition)
+        except constructor.ConstructorError as yex:
+            raise exc.WorkflowInspectionError([yex])
+        except Exception as ex:
+            raise ex
```

Comment:
Why reraise here? If you don't catch the exception, it will just propagate up the chain, and `instantiate()` will still be in the traceback.

```suggestion
```

Reply:
Refactored to fix another issue so this is no longer relevant.